File: api/LeadTracker.Core/Entities/Task.cs
using System.ComponentModel.DataAnnotations;

namespace LeadTracker.Core.Entities;

/// <summary>
/// Task entity - represents activities/tasks related to leads
/// </summary>
public class Task : TenantEntity
{
    [Required]
    [MaxLength(200)]
    public string Title { get; set; } = string.Empty;
    
    [MaxLength(2000)]
    public string? Description { get; set; }
    
    [Required]
    [MaxLength(20)]
    public string Type { get; set; } = "Call"; // Call, Email, Meeting, Follow-up, etc.
    
    [Required]
    [MaxLength(20)]
    public string Status { get; set; } = "Pending"; // Pending, Completed, Cancelled
    
    [Required]
    public DateTime DueDate { get; set; }
    
    public DateTime? CompletedAt { get; set; }
    
    [MaxLength(20)]
    public string Priority { get; set; } = "Medium"; // Low, Medium, High, Urgent
    
    [MaxLength(1000)]
    public string? Notes { get; set; }
    
    /// <summary>
    /// Duration in minutes (for completed tasks)
    /// </summary>
    public int? DurationMinutes { get; set; }
    
    // Reminder configuration
    public bool HasReminder { get; set; } = true;
    
    /// <summary>
    /// Exact time when reminder should be sent
    /// </summary>
    public DateTime? ReminderAt { get; set; }
    
    /// <summary>
    /// Number of minutes before DueDate to send reminder (default: 60 minutes)
    /// </summary>
    public int? ReminderMinutesBefore { get; set; } = 60;
    
    /// <summary>
    /// Whether reminder has been sent for this task
    /// </summary>
    public bool ReminderSent { get; set; } = false;
    
    // Recurrence configuration
    public bool IsRecurring { get; set; } = false;
    
    /// <summary>
    /// Recurrence pattern: "Daily", "Weekly", "Monthly", "Custom"
    /// </summary>
    [MaxLength(20)]
    public string? RecurrencePattern { get; set; }
    
    /// <summary>
    /// Interval for recurrence (e.g., every X days/weeks/months)
    /// </summary>
    public int? RecurrenceInterval { get; set; }
    
    /// <summary>
    /// When to stop creating recurring instances
    /// </summary>
    public DateTime? RecurrenceEndDate { get; set; }
    
    /// <summary>
    /// Reference to parent task if this is a recurring instance
    /// </summary>
    public Guid? ParentTaskId { get; set; }
    
    // Foreign Keys
    public Guid? LeadId { get; set; }
    
    [Required]
    public Guid AssignedUserId { get; set; }
    
    // Navigation properties
    public virtual Lead? Lead { get; set; }
    public virtual User AssignedUser { get; set; } = null!;
    public virtual Task? ParentTask { get; set; }
    public virtual ICollection<Task> RecurringInstances { get; set; } = new List<Task>();
    
    // Computed properties
    public bool IsOverdue => Status != "Completed" && Status != "Cancelled" && DueDate < DateTime.UtcNow;
    public bool IsCompleted => Status == "Completed";
    public bool IsToday
    {
        get
        {
            var now = DateTime.UtcNow;
            var todayStart = now.Date;
            var todayEnd = todayStart.AddDays(1);
            return DueDate >= todayStart && DueDate < todayEnd;
        }
    }
}

File: web/src/types/task.ts
export type TaskType = 'Call' | 'Email' | 'Meeting' | 'Follow-up' | 'Note' | 'Document';
export type TaskStatus = 'Pending' | 'Completed' | 'Cancelled';
export type TaskPriority = 'Low' | 'Medium' | 'High' | 'Urgent';

export interface Task {
  id: string;
  title: string;
  description?: string;
  type: TaskType;
  status: TaskStatus;
  dueDate: string;
  completedAt?: string;
  priority: TaskPriority;
  notes?: string;
  durationMinutes?: number;
  leadId?: string;
  leadTitle?: string; // Lead title for display
  assignedUserId: string; // Now required
  assignedUserName?: string; // String instead of user object
  organizationId: string;
  createdAt: string;
  updatedAt: string;
  createdBy?: string;
  updatedBy?: string;
  
  // Reminder configuration
  hasReminder: boolean;
  reminderAt?: string;
  reminderMinutesBefore?: number;
  reminderSent: boolean;
  
  // Recurrence configuration
  isRecurring: boolean;
  recurrencePattern?: 'Daily' | 'Weekly' | 'Monthly' | 'Custom';
  recurrenceInterval?: number;
  recurrenceEndDate?: string;
  parentTaskId?: string;
  
  // Computed properties
  isOverdue: boolean;
  isCompleted: boolean;
  isToday: boolean;
}

export interface CreateTaskDto {
  title: string;
  description?: string;
  type: TaskType;
  dueDate: string;
  priority?: TaskPriority;
  notes?: string;
  leadId?: string;
  assignedUserId: string; // Now required
  
  // Reminder configuration
  hasReminder?: boolean;
  reminderAt?: string;
  reminderMinutesBefore?: number;
  
  // Recurrence configuration
  isRecurring?: boolean;
  recurrencePattern?: 'Daily' | 'Weekly' | 'Monthly' | 'Custom';
  recurrenceInterval?: number;
  recurrenceEndDate?: string;
}

export interface UpdateTaskDto {
  title?: string;
  description?: string;
  type?: TaskType;
  status?: TaskStatus;
  dueDate?: string;
  priority?: TaskPriority;
  notes?: string;
  durationMinutes?: number;
  completedAt?: string | undefined;
  assignedUserId?: string;
  
  // Reminder configuration
  hasReminder?: boolean;
  reminderAt?: string;
  reminderMinutesBefore?: number;
  reminderSent?: boolean;
  
  // Recurrence configuration
  isRecurring?: boolean;
  recurrencePattern?: 'Daily' | 'Weekly' | 'Monthly' | 'Custom';
  recurrenceInterval?: number;
  recurrenceEndDate?: string;
  parentTaskId?: string;
}

export interface TaskQueryParams {
  page?: number;
  pageSize?: number;
  sortBy?: string;
  sortDirection?: 'asc' | 'desc';
  searchTerm?: string;
  leadId?: string;
  assignedUserId?: string;
  status?: TaskStatus;
  type?: TaskType;
  priority?: TaskPriority;
}

export interface PaginatedTasksResponse {
  data: Task[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

// Activity types for timeline
export interface Activity {
  id: string;
  type: 'task_completed' | 'task_created' | 'lead_status_changed' | 'lead_updated' | 'note_added' | 'call_made' | 'email_sent' | 'meeting_scheduled';
  title: string;
  description?: string;
  timestamp: string;
  userId?: string;
  userName?: string; // String instead of user object
  metadata?: {
    [key: string]: any;
  };
}

export interface ActivityTimeline {
  activities: Activity[];
  totalCount: number;
}

// Comment types
export interface Comment {
  id: string;
  content: string;
  leadId: string;
  userId: string;
  userName: string; // String instead of user object
  createdAt: string;
  updatedAt: string;
}

export interface CreateCommentDto {
  content: string;
  leadId: string;
}

export interface UpdateCommentDto {
  content: string;
}

// Attachment types
export interface Attachment {
  id: string;
  fileName: string;
  originalFileName: string;
  fileSize: number;
  mimeType: string;
  filePath: string;
  leadId: string;
  uploadedBy: string;
  uploadedByUserName: string; // String instead of user object
  createdAt: string;
}

export interface UploadAttachmentDto {
  file: File;
  leadId: string;
  description?: string;
}

// My Day view response
export interface MyDayTasksResponse {
  todayTasks: Task[];
  thisWeekTasks: Task[];
  overdueTasks: Task[];
  totalTodayCount: number;
  totalThisWeekCount: number;
  totalOverdueCount: number;
}
